// @revenium/perplexity/dist/models/Logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = exports.Logger = void 0;
const utils_1 = require("../utils");
class Logger {
    static debug(message, data) {
        if (this.shouldLog(utils_1.LOG_LEVELS[0])) {
            console.log(`[${utils_1.LOG_LEVELS[0]}] ${message}`, data || "");
        }
    }
    static info(message, data) {
        if (this.shouldLog(utils_1.LOG_LEVELS[1])) {
            console.log(`[${utils_1.LOG_LEVELS[1]}] ${message}`, data || "");
        }
    }
    static warning(message, data) {
        if (this.shouldLog(utils_1.LOG_LEVELS[2])) {
            console.warn(`[${utils_1.LOG_LEVELS[2]}] ${message}`, data || "");
        }
    }
    static error(message, data) {
        if (this.shouldLog(utils_1.LOG_LEVELS[3])) {
            console.error(`[${utils_1.LOG_LEVELS[3]}] ${message}`, data || "");
        }
    }
    static shouldLog(level) {
        const levels = utils_1.LOG_LEVELS;
        const currentLevel = levels.indexOf(this.logLevel.toUpperCase());
        const messageLevel = levels.indexOf(level);
        return messageLevel >= currentLevel;
    }
}
exports.Logger = Logger;
Logger.logLevel = process.env.REVENIUM_LOG_LEVEL || "INFO";
exports.logger = Logger;
